### DISCUSSION TITLE
My Diffbot lives!

### DISCUSSION THREAD

**Discussion by questrov** [fjp/2wd-robot#53] [Q&A] [answered]
Good news, I have a working Diffbot!  Close to SLAM but not quite there yet, details below...

Quick update for the record, after some pesky hardware issues I can now launch:
`roslaunch diffbot_bringup bringup_with_laser.launch model:=remo` on Pi
`roslaunch diffbot_slam diffbot_slam.launch` on WS
`roslaunch diffbot_bringup keyboard_teleop.launch` on WS
`roslaunch diffbot_bringup rpicamera.launch` on Pi
`rqt_image_view and/or add image panel in rviz` on WS

From above I can drive the Diffbot via keyboard and it moves as expected in rviz and real robot also have an image from the PiCamera V2.  Major milestone for this maker! 
![my_diffbot](https://user-images.githubusercontent.com/39848760/150626751-39f15d96-13be-4787-9840-02d037643dd2.jpeg)

Details of "yellow" messages and issues to date:

1.)  Diffbot_bringup throws,

`Using load_yaml() directly is deprecated. Use xacro.load_yaml() instead.` 

Have not investigated yet...

2.)  Diffbot_slam throws,

`[ WARN] [1642828211.397143360]: MessageFilter [target=odom ]: Dropped 100.00% of messages so far. Please turn the [ros.gmapping.message_filter] rosconsole logger to DEBUG for more information.` 

Some quick web searches turn up several threads having to do with transform issues.  Over my head at this point something I need to learn about.  Fuzzy idea of what a transform is/does but don't have details yet.

3.)  RVIZ has LaserScan Status: Error although I can `rostopic echo /diffbot/scan` and get plenty of 0's and 47's changing as I move my hand around the Lidar.  Think this may have to do with no. 2 above but not clear at this point.

4.) Pesky cause I keep forgetting but have to,

`/dev/ttyACM0 Permission denied` must `sudo chmod a+rw /dev/ttyACM0` after each reboot.  

Tried chmod 666 too and add user hasn't resolved it, kinda low on the list at this point.

5.) Never did get [rqt_robot_steering](http://wiki.ros.org/rqt_robot_steering) plugin to work on the real robot although it comes up automajically and works OK in the simulation with Gazebo.  Tried a few other cmd_vel topics such as `/diffbot/mobile_base_controller/cmd_vel` and `/diffbot/wheel_cmd_velocities` but they also give errors.  I'm launching with `rqt` as it's not coming up as advertised with rviz and ROS doesn't like it.  The error is,

`/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py:842: UserWarning: '/cmd_vel
' is not a legal ROS graph resource name. This may cause problems with other ROS tools
  super(Publisher, self).__init__(name, data_class, Registration.PUB)`  

Also low on the list maybe something to do with # 1 or 2 above as well?

6.)  Last but not least, that's a really cool BNO055 and usb-serial gizmo on the deck under the Pi UPS.  Works great stand alone with [bno055 ROS driver](http://wiki.ros.org/ros_imu_bno055) think I read it's not fully integrated yet?  Worth noting it can cause some USB port conflicts with the Lidar and originally changed USB ports in launch files back on the test bench.  Have it disconnected with the fresh install but looking forward to seeing the TF move around just because it's cool:-/

Still on the steep part of the ROS Diffbot learning curve but having fun now that all the breadboards are back in the drawer.  SLAM or bust!

**Comment by fjp** [comment, marked as answer]
Great work @questrov looks really good and nice progress! Thanks for sharing the update and your diffbot design :-)


> 1.) Diffbot_bringup throws,
>
> Using load_yaml() directly is deprecated. Use xacro.load_yaml() instead.

Thanks for reporting the yellow warning messages. I fixed 1.) for `diffbot_description` (83bcaf0) and `remo_description` (ros-mobile-robots/remo_description@67d27997). After you pull the latest changes for remo/diffbot (depending on which one you want to use further) these warnings should be gone.

> 2.) Diffbot_slam throws,
>
> [ WARN] [1642828211.397143360]: MessageFilter [target=odom ]: Dropped 100.00% of messages so far. Please turn the [ros.gmapping.message_filter] rosconsole logger to DEBUG for more information.
>
> 3.) RVIZ has LaserScan Status: Error although I can rostopic echo /diffbot/scan and get plenty of 0's and 47's changing as I move my hand around the Lidar. Think this may have to do with no. 2 above but not clear at this point.

Thanks for bringing this up again. This was already reported in #40 and tracked in #43 but not directly fixed on my side. It should be solved now with commit 60750e9ed2f0ecdb8f6250ca4c396238e8a0447a. There is now a new `rplidar.launch` that makes use of the `rplidar_ros` package from SLAMTEC. This new `rplidar.launch` is also being used now in the `bringup_with_laser.launch`. As you assumed correctly, this should solve both your issues 2.) and 3.)

> 4.) Pesky cause I keep forgetting but have to,
>
> /dev/ttyACM0 Permission denied must sudo chmod a+rw /dev/ttyACM0 after each reboot.
> 
> Tried chmod 666 too and add user hasn't resolved it, kinda low on the list at this point.

This shouldn't be required doing all the time. Have you added your user to the `dialout` group on the SBC (RPi)? It is explained here: https://ros-mobile-robots.com/hardware-interfaces/#usb-devices


> 5.) Never did get rqt_robot_steering plugin to work on the real robot although it comes up automajically and works OK in the simulation with Gazebo. Tried a few other cmd_vel topics such as /diffbot/mobile_base_controller/cmd_vel and /diffbot/wheel_cmd_velocities but they also give errors. I'm launching with rqt as it's not coming up as advertised with rviz and ROS doesn't like it. The error is,
>
> /opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py:842: UserWarning: '/cmd_vel ' is not a legal ROS graph resource name. This may cause problems with other ROS tools super(Publisher, self).__init__(name, data_class, Registration.PUB)
>
> Also low on the list maybe something to do with # 1 or 2 above as well?

This is strange. Have you tried `rosrun rqt_robot_steering rqt_robot_steering` in a new terminal and set the topic to `/diffbot/mobile_base_controller/cmd_vel`? Or which launch files are you trying to use?


> 6.) Last but not least, that's a really cool BNO055 and usb-serial gizmo on the deck under the Pi UPS. Works great stand alone with bno055 ROS driver think I read it's not fully integrated yet? Worth noting it can cause some USB port conflicts with the Lidar and originally changed USB ports in launch files back on the test bench. Have it disconnected with the fresh install but looking forward to seeing the TF move around just because it's cool:-/

Yes correct, the IMU is not yet supported but there is an open PR (#37) which provides good hints on how to get it working. This is still an open todo and I try to get to this after fixing some other open tasks with the MCU (for example more comfortable PID tuning on the low level PID mentioned in [#35](35#issuecomment-879229529) and #13 and now tracked in #54).

**Reply by fjp** [reply]
Please also note the change I just made in 3d412214ad0dc3d735764dc8c1e5e1948885f35e, switching the motor pins in the `diffbot_base_config.h`. In my case I needed to swap the pins because I missed doing that during a previous commit 2d99a0cf5dac95705a157c2595e39aa42667c0c4 where magic (hardcoded) numbers were used instead of those `#define`s.

You might need to swap the pin numbers back, otherwise you might get weird driving behavior after you pull the latest changes. Just a hint which could save you some debugging time :-)

**Reply by questrov** [reply]
Awesome, thanks Franz!  Yea, will have to change both motor and encoder pins as I’m using [Teensy 3.x Feather Adapter](https://www.adafruit.com/product/3200) and it doesn’t seem to break out Teensy pin 7 for the encoder.  Wound up changing the encoder 7-8 to 9-10 and motor pins to 1 & 2 as well to accommodate wire routing.

NOOB question, what is the proper way to “pull” the changes and updates you have made?  Have not done that before…

**Reply by fjp** [reply]
Usually you do a `git pull origin` to get the latest changes from the remote (named origin) but in case you made changes to your code base and committed them to your local git repo this will result in merge conflicts. Definitely backup the changes you made before pulling anything.

> I am assuming your remote is named origin. You can check the name with `git remote -v`

Without pulling anything you could first try to see the difference using

```
git fetch origin
git diff origin/noetic-devel
```

The first command might not be needed it simply checks the remote (origin) for changes but doesn't pull anything. The second command compares your local changes with the ones on the remote branch.

See also the difference between git fetch and git pull here https://stackoverflow.com/a/292359

If you have differences (which you know from the output of the `git diff` command) you could use vscode to merge the changes as needed (pick between your and my updates). For this to work however, you first have to commit your local changes to your local git repo. So that there are no local uncommitted or even untracked changes. You can check the status of that with `git status`.

**Reply by questrov** [reply]
Thanks that makes sense will give it a go.  Guess that brings up another item I've not gone thru the details yet, pushing files to github:-/  This is really the first project I've undertaken where it really needs to happen...

**Reply by questrov** [reply]
SLAM!  Thank you!  
![My Diffbot SLAM](https://user-images.githubusercontent.com/39848760/150650361-419dda97-fdb3-48ce-8512-dd1681e81698.png)

Managed to work out [Git Hub Pull Requests and Git Stash in Visual Studio Code](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github) with some help from [this write up](https://medium.com/@shrirammano/understanding-git-stash-and-using-it-easily-in-vscode-7da049160097) seemed to work and save my own tweaks locally in particular the encoder and motor pins for this build. 

All the other issues mentioned above resolved too!  Think the ACM0 permission thing just required reboot after adding user (duh) and robot steering works as advertised with `rosrun rqt_robot_steering rqt_robot_steering` added to my launch cheat sheet.

Time to dust off the big printer and make some REMO parts now that I have the STL files:-/

**Reply by fjp** [reply]
Great work @questrov! And by the way thanks for obtaining the REMO STL files. If you need any other support (also with the STL parts or 3D printing them) just let me know :-)

**Comment by ashwanthrkumar** [comment]
Hello, 
Could you share the project directory structure after catkin_ws
